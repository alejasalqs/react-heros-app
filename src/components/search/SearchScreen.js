import React, { useMemo } from "react";
import { useLocation } from "react-router";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import queryString from "query-string"; // -> npm i query-string
import { getHerosByName } from "../../selectos/getHerosByName";

export const SearchScreen = ({ history }) => {
  //hook de react router para obtener el location
  const location = useLocation();

  // Como es dificil trabajar los queryparams instalamos esta libreria para facilitar
  // regresa un obj con los queryparams que existan, si no hay params es undefined
  // para evitar errores se asigna un string vacio
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    search: q, // definimos el valor del query como defecto por si refrescan el navegador
  });

  const { search } = formValues;

  const heroesFilter = useMemo(() => getHerosByName(search), [q]);

  const handleSubmit = (e) => {
    e.preventDefault();
    // con este push ponemos lo del form de busqueda como query param, sin navegar a otro componente
    history.push(`?q=${search}`);
  };

  return (
    <>
      <h1>Search:</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search form</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              placeholder="Search"
              name="search"
              onChange={handleInputChange}
              value={search}
            />
            <button type="submit" className="btn btn-block btn-outline-primary">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results:</h4>
          <hr />
          {heroesFilter.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
