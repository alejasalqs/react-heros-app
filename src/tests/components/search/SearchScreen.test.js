import React from "react";
import { shallow } from "enzyme";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router";

describe("Pruebas en el componente <SearchScreen />", () => {
  const historyMock = {
    replace: jest.fn(),
  };
  const authContext = {
    dispatch: jest.fn(),
  };

  test("Debe de mostrarse correctamente ", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe mostrar a batman y el input con el valor del queryString ", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper.find("HeroCard").exists()).toBe(true);
  });
});
