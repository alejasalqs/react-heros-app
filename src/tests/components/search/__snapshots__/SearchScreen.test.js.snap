// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrarse con los valores por defecto  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "an3rmh",
          "pathname": "/search",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "an3rmh",
        "pathname": "/search",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Route
    component={[Function]}
    path="/search"
  />
</Router>
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrarse correctamente  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "q7oeeb",
          "pathname": "/search",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "q7oeeb",
        "pathname": "/search",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Route
    component={[Function]}
    path="/search"
  />
</Router>
`;
